### ISSUE TITLE
Ensure that api package doesn't expose impl classes

### ISSUE THREAD

**Issue by maskit** [apache/pulsar#1326]
#### Expected behavior

APIs shouldn't expose impl classes.

#### Actual behavior

```
$ git grep import | grep api/ | grep -v test/ | grep impl
pulsar-broker/src/main/java/org/apache/pulsar/client/api/RawReader.java:import org.apache.pulsar.client.impl.PulsarClientImpl;
pulsar-broker/src/main/java/org/apache/pulsar/client/api/RawReader.java:import org.apache.pulsar.client.impl.RawReaderImpl;
pulsar-client/src/main/java/org/apache/pulsar/client/api/AuthenticationFactory.java:import org.apache.pulsar.client.impl.auth.AuthenticationDisabled;
pulsar-client/src/main/java/org/apache/pulsar/client/api/ClientConfiguration.java:import org.apache.pulsar.client.impl.conf.ClientConfigurationData;
pulsar-client/src/main/java/org/apache/pulsar/client/api/Consumer.java:import org.apache.pulsar.client.impl.ConsumerStats;
pulsar-client/src/main/java/org/apache/pulsar/client/api/ConsumerConfiguration.java:import org.apache.pulsar.client.impl.conf.ConsumerConfigurationData;
pulsar-client/src/main/java/org/apache/pulsar/client/api/MessageBuilder.java:import org.apache.pulsar.client.impl.MessageBuilderImpl;
pulsar-client/src/main/java/org/apache/pulsar/client/api/MessageId.java:import org.apache.pulsar.client.impl.MessageIdImpl;
pulsar-client/src/main/java/org/apache/pulsar/client/api/Producer.java:import org.apache.pulsar.client.impl.ProducerStats;
pulsar-client/src/main/java/org/apache/pulsar/client/api/ProducerConfiguration.java:import org.apache.pulsar.client.impl.conf.ProducerConfigurationData;
pulsar-client/src/main/java/org/apache/pulsar/client/api/PulsarClient.java:import org.apache.pulsar.client.impl.ClientBuilderImpl;
pulsar-client/src/main/java/org/apache/pulsar/client/api/PulsarClient.java:import org.apache.pulsar.client.impl.PulsarClientImpl;
pulsar-client/src/main/java/org/apache/pulsar/client/api/ReaderConfiguration.java:import org.apache.pulsar.client.impl.conf.ReaderConfigurationData;
```

### Note

- Some will be fixed with #1321
- Some might be false alert since I just did grep `import` line
- Ideally api package shouldn't depend on any impl classes, but there might be acceptable cases

**Comment by merlimat** [comment]
This was fixed by placing the client API in its own separated module. Back-links from API to Impl are now forced to be explicits (eg: through reflection)

**Comment by merlimat** [comment]
(the fix was in #3309)
